File: TradeReplayAssistantAndroid/app/src/main/java/com/punglegod/trade/domain/ReportService.kt
package com.punglegod.trade.domain

class ReportService {
    fun generateSummary(
        trades: List<TradeSnapshot>,
        window: StatsRangeWindow,
        referenceMillis: Long
    ): TradeSummary {
        val start = window.startMillis(referenceMillis)
        val filtered = trades
            .asSequence()
            .filter { !it.isDeleted }
            .filter { it.closeTimeMillis in start..referenceMillis }
            .sortedBy { it.closeTimeMillis }
            .toList()

        if (filtered.isEmpty()) {
            return TradeSummary.Empty
        }

        val wins = filtered.filter { it.pnl > 0 }
        val losses = filtered.filter { it.pnl < 0 }

        val totalPnL = filtered.sumOf { it.pnl }
        val avgWin = if (wins.isEmpty()) 0.0 else wins.sumOf { it.pnl } / wins.size
        val avgLoss = if (losses.isEmpty()) 0.0 else losses.sumOf { kotlin.math.abs(it.pnl) } / losses.size

        val totalProfit = wins.sumOf { it.pnl }
        val totalLoss = losses.sumOf { kotlin.math.abs(it.pnl) }

        var running = 0.0
        var peak = 0.0
        var maxDrawdown = 0.0
        filtered.forEach {
            running += it.pnl
            peak = kotlin.math.max(peak, running)
            maxDrawdown = kotlin.math.max(maxDrawdown, peak - running)
        }

        return TradeSummary(
            totalTrades = filtered.size,
            winRate = wins.size.toDouble() / filtered.size.toDouble(),
            totalPnL = totalPnL,
            averageWin = avgWin,
            averageLoss = avgLoss,
            profitFactor = if (totalLoss == 0.0) totalProfit else totalProfit / totalLoss,
            maxDrawdown = maxDrawdown
        )
    }

    fun generateWeeklyReport(
        trades: List<TradeSnapshot>,
        referenceMillis: Long
    ): WeeklyReport {
        val dayMillis = 24L * 60 * 60 * 1_000
        val weekStart = referenceMillis - (6 * dayMillis)
        val weekEnd = referenceMillis

        val summary = generateSummary(
            trades = trades,
            window = StatsRangeWindow.LAST_7_DAYS,
            referenceMillis = referenceMillis
        )

        if (summary.totalTrades == 0) {
            return WeeklyReport.empty(referenceMillis)
        }

        val weekTrades = trades
            .filter { !it.isDeleted }
            .filter { it.closeTimeMillis in weekStart..weekEnd }

        val goodPractices = weekTrades
            .filter { it.pnl > 0 }
            .map { it.strategyTag }
            .toLabeledTop3(
                fallback = listOf("执行了既定止盈规则", "保持了仓位控制", "记录了复盘笔记")
            )

        val improvementAreas = weekTrades
            .filter { it.pnl < 0 }
            .map { if (it.mistakeTag.isBlank()) "出场纪律" else it.mistakeTag }
            .toLabeledTop3(
                fallback = listOf("减少冲动开仓", "严格执行止损", "收敛交易频率")
            )

        val nextActions = improvementAreas.map { "针对${it}设定明确触发条件并盘后打分" }

        return WeeklyReport(
            weekStartMillis = weekStart,
            weekEndMillis = weekEnd,
            summary = summary,
            goodPractices = goodPractices,
            improvementAreas = improvementAreas,
            nextWeekActions = nextActions
        )
    }
}

File: TradeReplayAssistant/Services/ReportService.swift
import Foundation

struct ReportService {
    func generateSummary(
        trades: [TradeSnapshot],
        within interval: DateInterval
    ) -> TradeSummary {
        let filtered = trades
            .filter { !$0.isDeleted }
            .filter { interval.contains($0.closeTime) }
            .sorted { $0.closeTime < $1.closeTime }

        guard !filtered.isEmpty else {
            return .empty
        }

        let wins = filtered.filter { $0.pnl > 0 }
        let losses = filtered.filter { $0.pnl < 0 }

        let totalPnL = filtered.reduce(0) { $0 + $1.pnl }
        let averageWin = wins.isEmpty ? 0 : wins.reduce(0) { $0 + $1.pnl } / Double(wins.count)
        let averageLoss = losses.isEmpty ? 0 : losses.reduce(0) { $0 + abs($1.pnl) } / Double(losses.count)
        let totalProfit = wins.reduce(0) { $0 + $1.pnl }
        let totalLoss = losses.reduce(0) { $0 + abs($1.pnl) }

        let runningEquity = filtered.reduce(into: [Double]()) { partial, trade in
            let next = (partial.last ?? 0) + trade.pnl
            partial.append(next)
        }

        var peak: Double = runningEquity.first ?? 0
        var maxDrawdown: Double = 0

        for equity in runningEquity {
            peak = max(peak, equity)
            maxDrawdown = max(maxDrawdown, peak - equity)
        }

        return TradeSummary(
            totalTrades: filtered.count,
            winRate: Double(wins.count) / Double(filtered.count),
            totalPnL: totalPnL,
            averageWin: averageWin,
            averageLoss: averageLoss,
            profitFactor: totalLoss == 0 ? totalProfit : totalProfit / totalLoss,
            maxDrawdown: maxDrawdown
        )
    }

    func generateWeeklyReport(
        trades: [TradeSnapshot],
        referenceDate: Date,
        calendar: Calendar = .current
    ) -> WeeklyReport {
        let weekInterval = weekInterval(for: referenceDate, calendar: calendar)
        let summary = generateSummary(trades: trades, within: weekInterval)

        guard summary.totalTrades > 0 else {
            return .empty(for: weekInterval)
        }

        let weekTrades = trades
            .filter { !$0.isDeleted }
            .filter { weekInterval.contains($0.closeTime) }

        let goodPractices = topLabels(
            from: weekTrades.filter { $0.pnl > 0 }.map(\.strategyTag),
            fallback: ["执行了既定止盈规则", "保持了仓位控制", "记录了复盘笔记"]
        )

        let improvementSources = weekTrades
            .filter { $0.pnl < 0 }
            .map { $0.mistakeTag.isEmpty ? "出场纪律" : $0.mistakeTag }

        let improvementAreas = topLabels(
            from: improvementSources,
            fallback: ["减少冲动开仓", "严格执行止损", "收敛交易频率"]
        )

        let nextWeekActions = improvementAreas.map { "针对\($0)设定明确触发条件并盘后打分" }

        return WeeklyReport(
            weekInterval: weekInterval,
            summary: summary,
            goodPractices: goodPractices,
            improvementAreas: improvementAreas,
            nextWeekActions: nextWeekActions
        )
    }

    private func weekInterval(for date: Date, calendar: Calendar) -> DateInterval {
        let components = calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: date)
        let start = calendar.date(from: components) ?? date
        let end = calendar.date(byAdding: .day, value: 7, to: start) ?? date
        return DateInterval(start: start, end: end)
    }

    private func topLabels(from values: [String], fallback: [String]) -> [String] {
        let cleaned = values
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }

        if cleaned.isEmpty {
            return fallback
        }

        let counter = cleaned.reduce(into: [String: Int]()) { partial, value in
            partial[value, default: 0] += 1
        }

        return counter
            .sorted {
                if $0.value == $1.value {
                    return $0.key < $1.key
                }
                return $0.value > $1.value
            }
            .prefix(3)
            .map(\.key)
    }
}
